### ISSUE TITLE
jenkins build spotbugs dependency slf4j slf4j-api vulnerability

### ISSUE THREAD

**Issue by datamaskin** [spotbugs/spotbugs#826]
Using the spotbugs plugin in my pom and deploying to Jenkins to build subsequently fails due to the slf4j slf4j-api vulnerability:
`/jenkins/jobs/workspace/target/my.jar/META-INF/maven/org.slf4j/slf4j-api/pom.xml , CVE-2018-8088 , Severity: High

org.slf4j.ext.EventData in the slf4j-ext module in QOS.CH SLF4J before 1.8.0-beta2 allows remote attackers to bypass intended access restrictions via crafted data.

The dependency is org.slf4j:slf4j-api:1.8.0-beta2 found in com.github.spotbugs:3.1.9`

**Comment by KengoTODA** [comment]
We need newer release from slf4j. https://www.slf4j.org/download.html

**Comment by ThrawnCA** [comment]
Can't we downgrade to 1.7.25?

**Comment by KengoTODA** [comment]
Possible but it won't work with Jigsaw

**Comment by kebernet** [comment]
Is it possible to just use j.u.l here? Honestly the SLF4J is a giant PITA to as part of Gradle 5 builds with Spring Boot because you end up with the "Logger Loop" between L4J and SLF4J. 

Is there critical functionality that SLF4J is providing here?

**Comment by JLLeitschuh** [comment]
Depending upon how the spotbugs Gradle plugin SLF4J might be creating a bridge between the Gradle logger and spotbugs, thus allowing the logging in Spotbugs to be handled by the Gradle logger. I'm not certain though.

**Comment by ThrawnCA** [comment]
There has been some discussion (#635) around whether to use SLF4J or stick with JUL. The consensus was that SLF4J is much nicer to work with, but if you're running into concrete cases where the facade causes difficulty, we can potentially revisit that.

**Comment by Vampire** [comment]
@KengoTODA what problems do you see with slf4j 1.7.25 and Jigsaw?
I successfully used this combination without any problem as far as I could see.
I'm mainly curious, I wondered that you depend on a beta-version and looked around, finding this issue so thought I chime in here.

**Comment by KengoTODA** [comment]
I had misunderstanding about this issue: SLF4J **before** 1.8.0-beta2 is vulnerable so current SpotBugs in `master` branch should have no vulnerability. I will close this issue.

**Comment by KengoTODA** [comment]
@Vampire what I know is here: 

* https://www.slf4j.org/faq.html#changesInVersion18
* https://jira.qos.ch/browse/SLF4J-372

In theory, all SLF4J bindings has the same package `org.slf4j.impl` and it's not allowed by jigsaw. I think it has no problem when we have only one binding in classpath, but at least in corner cases, SLF4J 1.7 leads to non intuitive trouble. It's my opinion.

Yes it's beta version but it already has [many usages](https://mvnrepository.com/artifact/org.slf4j/slf4j-api/1.8.0-beta2) so I think it's not problem. I just worry the activity of SLF4J project, but maybe no problem... I know it does not react so quickly...

**Comment by KengoTODA** [comment]
https://jira.qos.ch/plugins/servlet/mobile#issue/SLF4J-451
The issue is still not closed, I will reopen tgis issue.

**Comment by KengoTODA** [comment]
I'll re-close this issue. Its points are:

1. According to [this report](https://nvd.nist.gov/vuln/detail/CVE-2018-8088), vulnerability exist not in slf4j-api but slf4j-ext module.
2. Its version 1.7.x is also vulnerable.
3. [slf4j-api does not depends on slf4j-ext](https://mvnrepository.com/artifact/org.slf4j/slf4j-api/1.7.25).

So removing usage on slf4j-ext is fully users' responsibility. And SpotBugs is not get affected by this vulnerability.

@datamaskin I'm not sure why your pom of slf4j-api contains reference to slf4j-ext. Is it still reproducible? If possible, consider to share your `my.jar/META-INF/maven/org.slf4j/slf4j-api/pom.xml`.

**Comment by Vampire** [comment]
> In theory, all SLF4J bindings has the same package `org.slf4j.impl` and it's not allowed by jigsaw. I think it has no problem when we have only one binding in classpath, but at least in corner cases, SLF4J 1.7 leads to non intuitive trouble. It's my opinion.

Well, having multiple SLF4J bindings in the classpath is a soft error anyway and will cause a big warning.
Imho it is good that Jigsaw prevents it and you can fix your dependencies, especially as warnings tend to be ignored easily.

> Yes it's beta version but it already has [many usages](https://mvnrepository.com/artifact/org.slf4j/slf4j-api/1.8.0-beta2) so I think it's not problem. I just worry the activity of SLF4J project, but maybe no problem... I know it does not react so quickly...

Well, just because others do things wrongly, doesn't mean you have to replicate that behavior. :-D
Imho it is a no-go to use SNAPSHOT versions, beta versions, and similar in production.
But that's up to you guys of course. :-)

Also not everyone uses 1.8 intentionally.
log4j-slf4j-impl for example depended upon 1.8 accidentally and anyone using it thus also used 1.8 without asking.
Until I asked them and it came out that it was by accident and they split it into log4j-slf4j-impl and log4j-slf4j18-impl (which is not just because of the beta-state, but also because separate artifacts are necessary there). My point is just that the dependency was by accident.

**Comment by Vampire** [comment]
Acutally this also seems to be a problem when you run SpotBugs using the Gradle plugin:

```log
SLF4J: No SLF4J providers were found.
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#noProviders for further details.
SLF4J: Class path contains SLF4J bindings targeting slf4j-api versions prior to 1.8.
SLF4J: Ignoring binding found at [jar:file:/D:/Dateien/.gradle/wrapper/dists/gradle-4.10.2-bin/cghg6c4gf4vkiutgsab8yrnwv/gradle-4.10.2/lib/gradle-logging-4.10.2.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#ignoredBindings for an explanation.
```

**Comment by chwba** [comment]
So how could one run SpotBugs if really needed. I'm using the plugin and getting this as well.

**Comment by Vampire** [comment]
And your problem is? This is just a warning.

**Comment by chwba** [comment]
Theres no output whatsoever, only this "warning".

**Comment by ThrawnCA** [comment]
What plugin is giving this error? Is it the OWASP dependency checker?

Since SpotBugs is not using `slf4j-ext`, this bug should probably be filed against whatever plugin is reporting it.

**Comment by flexanji** [comment]
Hey, I am using spotbugs 3.1.12 and log4j 1.2.16 with gradle 5.6.4. My logs are not printing due to conflict between spotbugs and my code.
SLF4J: No SLF4J providers were found.
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#noProviders for further details.
SLF4J: Class path contains SLF4J bindings targeting slf4j-api versions prior to 1.8.
SLF4J: Ignoring binding found at [jar:file:xxxx/yyyyy/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#ignoredBindings for an explanation.

the problem is that gradle spotbugs plugin adds slf4-simple dependency with fixed version 1.8.0-beta4 https://github.com/spotbugs/spotbugs-gradle-plugin/blob/master/build.gradle#L28
